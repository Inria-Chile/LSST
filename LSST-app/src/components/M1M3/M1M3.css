.mirror-container{
    display: inline-block;
}