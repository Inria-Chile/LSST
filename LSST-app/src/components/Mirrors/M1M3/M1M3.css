.mirror-container{
    display: inline-block;
}

.circle-overlay {
    stroke: #2d3d4e;
    stroke-width: 10px;
}