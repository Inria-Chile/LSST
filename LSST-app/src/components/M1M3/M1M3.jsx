import React, { Component } from 'react'
import './M1M3.css';
import { checkStatus, parseJSON } from "../Utils/Utils"

class M1M3 extends Component {

  constructor(props){
    super(props);
    this.data = [];
    this.state = {
        data: [],
        xRadius: 0,
        yRadius: 0,
        maxRadius: 0
    }
    fetch('m1m3/actuators', {
        accept: "application/json"
    })
    .then(checkStatus)
    .then(parseJSON)
    .then((data) => {
        let yMax = -Infinity;
        let xMax = -Infinity;
        let yMin = Infinity;
        let xMin = Infinity;
        let maxRadius = 0;
        data.results.forEach(act => {
            if(xMax < act.position[0])
                xMax = act.position[0];
            if(xMin > act.position[0])
                xMin = act.position[0];
            if(yMax < act.position[1])
                yMax = act.position[1];
            if(yMin > act.position[1])
                yMin = act.position[1];
            if(maxRadius < Math.sqrt(Math.pow(act.position[0], 2) + Math.pow(act.position[1], 2)))
                maxRadius = Math.sqrt(Math.pow(act.position[0], 2) + Math.pow(act.position[1], 2));
        });
        this.setState({
            data: data.results,
            xRadius: (xMax - xMin)/2,
            yRadius: (yMax - yMin)/2,
            maxRadius: maxRadius
        })
    });
  }
  
  componentDidMount() {
  }

  componentDidUpdate(){
  }

  
  render() {
    return (
        <div className="mirror-container" ref="container">
            <svg
                className="svg-container"
                height={this.props.height}
                width={this.props.width}>
                <circle
                    cx={this.state.xRadius*this.props.scale+this.props.margin}
                    cy={this.state.yRadius*this.props.scale+this.props.margin}
                    key={'dsadsadsa'}
                    fill={'#04070a'}
                    r={this.state.maxRadius*this.props.scale*1.05}/>
                <g className="scatter">
                    {this.state.data.map(act => (
                        <circle
                            cx={(act.position[0] + this.state.xRadius)*this.props.scale+this.props.margin}
                            cy={(act.position[1] + this.state.yRadius)*this.props.scale+this.props.margin}
                            key={act.actuatorID}
                            fill={this.props.colormap(Math.sqrt(Math.pow(act.position[0], 2) + Math.pow(act.position[1], 2))/this.state.maxRadius)}
                            r={8}/>
                        )
                    )}
                </g>
            </svg>
        </div>
    );
  }
}


export default M1M3;