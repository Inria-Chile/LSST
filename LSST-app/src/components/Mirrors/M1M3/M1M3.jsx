import React, { Component } from 'react'
import './M1M3.css';
import { checkStatus, parseJSON } from "../../Utils/Utils"
import * as d3 from 'd3';

class M1M3 extends Component {

  constructor(props){
    super(props);
    this.data = [];
    this.state = {
        data: [],
        xRadius: 0,
        yRadius: 0,
        maxRadius: 0
    }
    fetch('backend/m1m3/actuators', {
        accept: "application/json"
    })
    .then(checkStatus)
    .then(parseJSON)
    .then((data) => {
        let yMax = -Infinity;
        let xMax = -Infinity;
        let yMin = Infinity;
        let xMin = Infinity;
        let maxRadius = 0;
        data.results.forEach(act => {
            if(xMax < act.position[0])
                xMax = act.position[0];
            if(xMin > act.position[0])
                xMin = act.position[0];
            if(yMax < act.position[1])
                yMax = act.position[1];
            if(yMin > act.position[1])
                yMin = act.position[1];
            if(maxRadius < Math.sqrt(Math.pow(act.position[0], 2) + Math.pow(act.position[1], 2)))
                maxRadius = Math.sqrt(Math.pow(act.position[0], 2) + Math.pow(act.position[1], 2));
        });
        this.setState({
            data: data.results,
            xRadius: (xMax - xMin)/2,
            yRadius: (yMax - yMin)/2,
            maxRadius: maxRadius
        })
    });
  }
  
    zoomed = () => {
        let xRadius = this.state.xRadius;
        let yRadius = this.state.yRadius;
        let scale = Math.max(this.state.xRadius, this.state.yRadius)*this.props.width/65000;
        let id = this.props.id;
        d3.event.transform.x = Math.min(0, Math.max(d3.event.transform.x, 2*xRadius*scale - 2*xRadius*scale * d3.event.transform.k));
        d3.event.transform.y = Math.min(0, Math.max(d3.event.transform.y, 2*yRadius*scale - 2*yRadius*scale * d3.event.transform.k));
        d3.select('#scatter-'+id).attr("transform", d3.event.transform);
        d3.select('#background-circle-'+id).attr("transform", d3.event.transform);
    }

    componentDidUpdate(prevProps, prevState){
        d3.select('#circle-overlay-'+this.props.id).call(d3.zoom().scaleExtent([1, Infinity]).on("zoom", this.zoomed));
    }

    componentDidMount() {
    }

    
    render() {
        let scale = Math.max(this.state.xRadius, this.state.yRadius)*this.props.width/65000;
        return (
            <div className="mirror-container" ref="container">
                <svg
                    className="svg-container"
                    height={this.props.height + "px"}
                    width={this.props.width + "px"}>
                    <circle id={"background-circle-"+this.props.id} className="circle-overlay"
                        cx={this.state.xRadius*scale+this.props.margin}
                        cy={this.state.yRadius*scale+this.props.margin}
                        key={'background'}
                        fill={'#04070a'}
                        r={this.state.maxRadius*scale*1.15}
                        pointerEvents="all"
                        />
                    <g id={"scatter-"+this.props.id} className="scatter">
                        {this.state.data.map(act => (
                            <circle
                                cx={(act.position[0] + this.state.xRadius)*scale+this.props.margin}
                                cy={(act.position[1] + this.state.yRadius)*scale+this.props.margin}
                                key={act.actuatorID}
                                fill={this.props.colormap(Math.sqrt(Math.pow(act.position[0], 2) + Math.pow(act.position[1], 2))/this.state.maxRadius)}
                                r={this.state.maxRadius*scale/21}/>
                            )
                        )}
                    </g>
                    <circle id={"circle-overlay-"+this.props.id}
                        cx={this.state.xRadius*scale+this.props.margin}
                        cy={this.state.yRadius*scale+this.props.margin}
                        key={'overlay'}
                        fill={'none'}
                        r={this.state.maxRadius*scale*1.15}
                        pointerEvents="all"
                        />
                </svg>
            </div>
        );
    }
}


export default M1M3;