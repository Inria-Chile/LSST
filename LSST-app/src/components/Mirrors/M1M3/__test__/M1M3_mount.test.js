import React from 'react';
import M1M3 from '../M1M3';
import Enzyme from 'enzyme';
import Adapter from 'enzyme-adapter-react-16';
import * as d3 from 'd3';
import {mount} from "enzyme"; 
import sinon from 'sinon';
import toJson from 'enzyme-to-json';
Enzyme.configure({ adapter: new Adapter() });

global.fetch = require('jest-fetch-mock');

describe('Update state test',function(){
    let mountedM1M3,spy;
    let mirrorSize = 250;
    let mirrorMargin = 30;
    let colormap1 = d3.scaleSequential((t) => d3.hsl(360, 1.0-t*t*0.1, 0.12+t*t*0.58) + "");
    let colormap2 = d3.scaleSequential((t) => d3.hsl(270, 0.9+t*t*0.1, 0.7-t*t*0.58) + "");
    let colormap3 = d3.scaleSequential((t) => d3.hsl(140, 0.9+t*t*0.1, 0.7-t*t*0.58) + "");
    const M1M3Component = ()=>{
      if(!mountedM1M3){
        mountedM1M3 = mount(
          <M1M3 width={mirrorSize} height={mirrorSize} id="m3" margin={mirrorMargin} colormap={colormap1}/>
        );
      }
      return mountedM1M3;
    };

    beforeEach(()=>{
      mountedM1M3 = undefined;

    
        fetch.resetMocks();
        const data = {
        results : []
        }
    
        const primero = {
        position : [3,4]
        }
    
        const segundo = {
        position : [5,6]
        }
    
        data.results.push(primero);
        data.results.push(segundo);
        fetch.mockResponse(JSON.stringify(data));  
    });

    afterEach(()=>{
      spy.restore();
    });

    it('should call component did update when setProps',()=>{
      spy = sinon.spy(M1M3.prototype,'componentDidUpdate');
      const tree = M1M3Component();
      expect(spy.calledOnce).toEqual(false);
      tree.setState({xRadius : 2});
      expect(spy.calledOnce).toEqual(true);
    });

    it('should create the right snapshot',()=>{
      spy = sinon.spy(M1M3.prototype,'componentDidUpdate');
      let tree = M1M3Component();
      expect(spy.calledOnce).toEqual(false);
      tree.setState({xRadius : 5});
      tree = M1M3Component();
      expect(toJson(tree)).toMatchSnapshot();
    });

  });  