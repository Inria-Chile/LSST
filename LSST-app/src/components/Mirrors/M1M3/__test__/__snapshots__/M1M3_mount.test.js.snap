// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Update state test should create the right snapshot 1`] = `
<M1M3
  colormap={[Function]}
  height={250}
  id="m3"
  margin={30}
  width={250}
>
  <div
    className="mirror-container"
  >
    <svg
      className="svg-container"
      height="250px"
      width="250px"
    >
      <circle
        className="circle-overlay"
        cx={30.096153846153847}
        cy={30}
        fill="#04070a"
        id="background-circle-m3"
        key="background"
        pointerEvents="all"
        r={0}
      />
      <g
        className="scatter"
        id="scatter-m3"
      />
      <circle
        cx={30.096153846153847}
        cy={30}
        fill="none"
        id="circle-overlay-m3"
        key="overlay"
        pointerEvents="all"
        r={0}
      />
    </svg>
  </div>
</M1M3>
`;
